larpandoracontent/LArVertex/VertexRefinementAlgorithm.cc

```diff
@@ -0,0 +1,246 @@
+        parameters.m_position = position3D;
+        parameters.m_vertexLabel = VERTEX_INTERACTION;
+        parameters.m_vertexType = VERTEX_3D;
+
+        const Vertex *pNewVertex(NULL);
+        PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::Vertex::Create(*this, parameters, pNewVertex));
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+CartesianVector VertexRefinementAlgorithm::RefineVertexTwoD(const ClusterList &clusterList, const CartesianVector &originalVtxPos) const
+{
+    CartesianPointVector intercepts, directions;
+    FloatVector weights;
+
+    for (const Cluster *const pCluster : clusterList)
+    {
+        if (LArClusterHelper::GetClosestDistance(originalVtxPos, pCluster) > 10)
+            continue;
+
+        if (pCluster->GetNCaloHits() < m_minimumHitsCut)
+            continue;
+
+        CartesianVector centroid(-999, -999, -999);
+        LArPcaHelper::EigenValues eigenValues(-999, -999, -999);
+        LArPcaHelper::EigenVectors eigenVectors;
+        CartesianPointVector pointVector;
+
+        LArClusterHelper::GetCoordinateVector(pCluster, pointVector);
+        LArPcaHelper::RunPca(pointVector, centroid, eigenValues, eigenVectors);
+
+        intercepts.push_back(LArClusterHelper::GetClosestPosition(originalVtxPos, pCluster));
+        directions.push_back(eigenVectors.at(0).GetUnitVector());
+        weights.push_back(1.f / ((LArClusterHelper::GetClosestPosition(originalVtxPos, pCluster) - originalVtxPos).GetMagnitudeSquared() + 1));
+    }
+
+    CartesianVector newVtxPos(originalVtxPos);
+    if (intercepts.size() > 1)
+        GetBestFitPoint(intercepts, directions, weights, newVtxPos);
+
+    if ((newVtxPos - originalVtxPos).GetMagnitude() > m_twoDDistanceCut)
+        return originalVtxPos;
+
+    return newVtxPos;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void VertexRefinementAlgorithm::GetBestFitPoint(const CartesianPointVector &intercepts, const CartesianPointVector &directions,
+    const FloatVector &weights, CartesianVector &bestFitPoint) const
+{
+    const int n(intercepts.size());
+
+    Eigen::VectorXd d = Eigen::VectorXd::Zero(3 * n);
+    Eigen::MatrixXd G = Eigen::MatrixXd::Zero(3 * n, n + 3);
+
+    for (int i = 0; i < n; ++i)
+    {
+        d(3 * i) = intercepts[i].GetX() * weights[i];
+        d(3 * i + 1) = intercepts[i].GetY() * weights[i];
+        d(3 * i + 2) = intercepts[i].GetZ() * weights[i];
+
+        G(3 * i, 0) = weights[i];
+        G(3 * i + 1, 1) = weights[i];
+        G(3 * i + 2, 2) = weights[i];
+
+        G(3 * i, i + 3) = -directions[i].GetX();
+        G(3 * i + 1, i + 3) = -directions[i].GetY();
+        G(3 * i + 2, i + 3) = -directions[i].GetZ();
+    }
+
+    Eigen::VectorXd m = (G.transpose() * G).inverse() * G.transpose() * d;
```

Comment:
Do we need to check that G.transpose() * G  is not singular? What's the behaviour of Eigen if it is?

Reply:
So the `inverse()` function returns a matrix of NaNs in the case of singular matrices. I think this was being implicitly caught by physics requirements on the position of the vertex later on but this should definitely be prevent explicitly so I've added a requirement on the determinant in my new commit. 

Note that this includes a vector initialised with -999.f values not 0.f. Despite the above conversation this is deliberate, I want to be absolutely sure that a return due to singular matrices will cause line 187 to fail. (0,0,0) is a physically "sensible" position in most detectors coordinate systems whereas (-999,-999,-999) shouldn't be.

Reply:
Is this latter scenario a reason to opt for initialising this particular vector with a set of std::numeric_limits<float>::max() values instead. I agree -999 x 3 is a very unlikely coordinate, but numeric limits should be safer still if we want to avoid doing this with exceptions.

Reply:
Yep that's a good shout!